File: server/src/services/extraction/schemas.ts
import { z } from 'zod/v4';
import type { DocClass } from '../classification/types';

// ---- Paystub ----
export const paystubSchema = z.object({
  employer_name: z.string(),
  employee_name: z.string().optional(),
  pay_period_start: z.string().optional(),
  pay_period_end: z.string().optional(),
  pay_date: z.string().optional(),
  pay_frequency: z.enum(['weekly', 'biweekly', 'semimonthly', 'monthly']).optional(),
  gross_pay: z.number(),
  federal_tax: z.number().optional(),
  state_tax: z.number().optional(),
  social_security: z.number().optional(),
  medicare: z.number().optional(),
  health_insurance: z.number().optional(),
  retirement_401k: z.number().optional(),
  other_deductions: z.array(z.object({ name: z.string(), amount: z.number() })).optional(),
  net_pay: z.number(),
  ytd_gross: z.number().optional(),
  ytd_net: z.number().optional(),
  hours_worked: z.number().optional(),
  hourly_rate: z.number().optional(),
});

// ---- Bank Statement ----
export const bankStatementSchema = z.object({
  institution_name: z.string(),
  account_type: z.enum(['checking', 'savings', 'investment']).optional(),
  account_number_last4: z.string().optional(),
  statement_period_start: z.string().optional(),
  statement_period_end: z.string().optional(),
  beginning_balance: z.number(),
  ending_balance: z.number(),
  total_deposits: z.number().optional(),
  total_withdrawals: z.number().optional(),
  transactions: z.array(z.object({
    date: z.string(),
    description: z.string(),
    amount: z.number(),
    type: z.enum(['credit', 'debit']),
  })).optional(),
});

// ---- W-2 ----
export const w2Schema = z.object({
  employer_name: z.string(),
  employer_ein: z.string().optional(),
  employee_name: z.string().optional(),
  employee_ssn_last4: z.string().optional(),
  tax_year: z.string().optional(),
  wages: z.number(),
  federal_tax_withheld: z.number(),
  social_security_wages: z.number().optional(),
  social_security_tax: z.number().optional(),
  medicare_wages: z.number().optional(),
  medicare_tax: z.number().optional(),
  state: z.string().optional(),
  state_wages: z.number().optional(),
  state_tax: z.number().optional(),
});

// ---- Tax Return ----
export const taxReturnSchema = z.object({
  tax_year: z.string(),
  return_type: z.enum(['federal', 'state']).optional(),
  filing_status: z.enum(['single', 'married_jointly', 'married_separately', 'head_of_household', 'qualifying_surviving_spouse']).optional(),
  adjusted_gross_income: z.number(),
  taxable_income: z.number().optional(),
  total_tax: z.number().optional(),
  total_payments: z.number().optional(),
  refund_amount: z.number().optional(),
  amount_owed: z.number().optional(),
});

// ---- Credit Card Statement ----
export const creditCardStatementSchema = z.object({
  issuer: z.string(),
  account_number_last4: z.string().optional(),
  statement_period_start: z.string().optional(),
  statement_period_end: z.string().optional(),
  previous_balance: z.number(),
  payments: z.number().optional(),
  new_charges: z.number().optional(),
  ending_balance: z.number(),
  minimum_payment_due: z.number().optional(),
  payment_due_date: z.string().optional(),
  credit_limit: z.number().optional(),
  available_credit: z.number().optional(),
  cash_advances: z.number().optional(),
});

// ---- Profit & Loss Statement ----
export const profitLossSchema = z.object({
  business_name: z.string(),
  gross_revenue: z.number(),
  net_profit: z.number(),
  period_start: z.string().optional(),
  period_end: z.string().optional(),
  total_expenses: z.number().optional(),
  owner_name: z.string().optional(),
  cost_of_goods_sold: z.number().optional(),
  payroll_expenses: z.number().optional(),
  rent_expense: z.number().optional(),
  utilities: z.number().optional(),
  other_expenses: z.array(z.object({ name: z.string(), amount: z.number() })).optional(),
});

// ---- Retirement Account ----
export const retirementAccountSchema = z.object({
  institution_name: z.string(),
  account_type: z.enum(['IRA', '401k', '403b', 'pension', 'other_retirement']),
  ending_balance: z.number(),
  account_number_last4: z.string().optional(),
  statement_period_end: z.string().optional(),
  account_holder_name: z.string().optional(),
  employer_name: z.string().optional(),
});

// ---- Collection Letter ----
export const collectionLetterSchema = z.object({
  collection_agency_name: z.string(),
  amount_claimed: z.number(),
  original_creditor: z.string().optional(),
  account_number_last4: z.string().optional(),
  letter_date: z.string().optional(),
  debt_type: z.string().optional(),
  references_lawsuit: z.boolean().optional(),
  references_judgment: z.boolean().optional(),
  judgment_amount: z.number().optional(),
  court_name: z.string().optional(),
  collection_agency_address: z.string().optional(),
  phone: z.string().optional(),
});

// ---- Legal Document ----
export const legalDocumentSchema = z.object({
  document_type: z.enum(['summons', 'complaint', 'judgment', 'garnishment_order', 'foreclosure_notice', 'other']),
  plaintiff_name: z.string(),
  defendant_name: z.string().optional(),
  case_number: z.string().optional(),
  court_name: z.string().optional(),
  court_address: z.string().optional(),
  filing_date: z.string().optional(),
  case_type: z.string().optional(),
  amount_claimed: z.number().optional(),
  judgment_amount: z.number().optional(),
  garnishment_amount: z.number().optional(),
  property_address: z.string().optional(),
});

// ---- Vehicle Loan Statement ----
export const vehicleLoanSchema = z.object({
  lender_name: z.string(),
  current_balance: z.number(),
  account_number_last4: z.string().optional(),
  interest_rate: z.number().optional(),
  monthly_payment: z.number().optional(),
  vehicle_description: z.string().optional(),
  loan_origination_date: z.string().optional(),
  payoff_amount: z.number().optional(),
  lender_address: z.string().optional(),
});

// ---- Mortgage Statement ----
export const mortgageSchema = z.object({
  lender_name: z.string(),
  current_balance: z.number(),
  loan_number: z.string().optional(),
  property_address: z.string().optional(),
  interest_rate: z.number().optional(),
  monthly_payment: z.number().optional(),
  statement_period_end: z.string().optional(),
  loan_type: z.enum(['first_mortgage', 'second_mortgage', 'heloc', 'other']).optional(),
  escrow_balance: z.number().optional(),
  payoff_amount: z.number().optional(),
  lender_address: z.string().optional(),
});

// ---- Social Security Letter ----
export const socialSecuritySchema = z.object({
  monthly_benefit: z.number(),
  benefit_type: z.enum(['SSDI', 'SSI', 'retirement', 'survivor', 'other']).optional(),
  effective_date: z.string().optional(),
  recipient_name: z.string().optional(),
  net_monthly_benefit: z.number().optional(),
  medicare_premium: z.number().optional(),
  annual_benefit: z.number().optional(),
});

// ---- ID Document (Driver's License) ----
export const idDocumentSchema = z.object({
  full_name: z.string(),
  date_of_birth: z.string().optional(),
  license_number: z.string().optional(),
  expiration_date: z.string().optional(),
  address: z.string().optional(),
  state: z.string().optional(),
  sex: z.string().optional(),
});

// ---- Social Security Card ----
export const socialSecurityCardSchema = z.object({
  full_name: z.string(),
  ssn_last4: z.string().optional(),
});

// ---- Brokerage Statement ----
export const brokerageStatementSchema = z.object({
  institution_name: z.string(),
  account_holder_name: z.string().optional(),
  account_number_last4: z.string().optional(),
  statement_period_start: z.string().optional(),
  statement_period_end: z.string().optional(),
  beginning_value: z.number(),
  ending_value: z.number(),
});

// ---- Vehicle Title ----
export const vehicleTitleSchema = z.object({
  title_number: z.string().optional(),
  vin: z.string().optional(),
  year: z.string().optional(),
  make: z.string().optional(),
  model: z.string().optional(),
  body_style: z.string().optional(),
  legal_owner_name: z.string().optional(),
  registered_owner_name: z.string().optional(),
  issue_date: z.string().optional(),
  odometer_miles: z.number().optional(),
  legal_owner_address: z.string().optional(),
  prior_title_state: z.string().optional(),
});

// ---- Tax 1099 ----
export const tax1099Schema = z.object({
  form_variant: z.string(),
  payer_name: z.string(),
  recipient_name: z.string().optional(),
  recipient_ssn_last4: z.string().optional(),
  tax_year: z.string().optional(),
  total_amount: z.number(),
  federal_tax_withheld: z.number().optional(),
});

// ---- Generic (for other/unclassified) ----
export const genericSchema = z.record(z.string(), z.unknown());

// ---- Inferred TypeScript types (source of truth: the Zod schemas above) ----
export type PaystubData = z.infer<typeof paystubSchema>;
export type BankStatementData = z.infer<typeof bankStatementSchema>;
export type W2Data = z.infer<typeof w2Schema>;
export type TaxReturnData = z.infer<typeof taxReturnSchema>;
export type CreditCardData = z.infer<typeof creditCardStatementSchema>;
export type ProfitLossData = z.infer<typeof profitLossSchema>;
export type RetirementAccountData = z.infer<typeof retirementAccountSchema>;
export type CollectionLetterData = z.infer<typeof collectionLetterSchema>;
export type LegalDocumentData = z.infer<typeof legalDocumentSchema>;
export type VehicleLoanData = z.infer<typeof vehicleLoanSchema>;
export type MortgageData = z.infer<typeof mortgageSchema>;
export type SocialSecurityData = z.infer<typeof socialSecuritySchema>;
export type IdDocumentData = z.infer<typeof idDocumentSchema>;
export type SocialSecurityCardData = z.infer<typeof socialSecurityCardSchema>;
export type BrokerageStatementData = z.infer<typeof brokerageStatementSchema>;
export type VehicleTitleData = z.infer<typeof vehicleTitleSchema>;
export type Tax1099Data = z.infer<typeof tax1099Schema>;
export type GenericData = z.infer<typeof genericSchema>;

export type ExtractionData =
  | PaystubData
  | BankStatementData
  | W2Data
  | TaxReturnData
  | CreditCardData
  | ProfitLossData
  | RetirementAccountData
  | CollectionLetterData
  | LegalDocumentData
  | VehicleLoanData
  | MortgageData
  | SocialSecurityData
  | IdDocumentData
  | SocialSecurityCardData
  | BrokerageStatementData
  | VehicleTitleData
  | Tax1099Data
  | GenericData;

// ---- Schema lookup by doc class ----
const schemaMap: Partial<Record<DocClass, z.ZodType>> = {
  'payStub.us': paystubSchema,
  'bankStatement.us.checking': bankStatementSchema,
  'bankStatement.us.savings': bankStatementSchema,
  'tax.us.1040': taxReturnSchema,
  'tax.us.w2': w2Schema,
  creditCard: creditCardStatementSchema,
  // Legacy retirement classes route to the unified schema
  ira_statement: retirementAccountSchema,
  '401k_statement': retirementAccountSchema,
  // New classes
  profit_loss_statement: profitLossSchema,
  retirement_account: retirementAccountSchema,
  collection_letter: collectionLetterSchema,
  legal_document: legalDocumentSchema,
  vehicle_loan_statement: vehicleLoanSchema,
  'mortgage.us': mortgageSchema,
  social_security_letter: socialSecuritySchema,
  idDocument: idDocumentSchema,
  social_security_card: socialSecurityCardSchema,
  brokerage_statement: brokerageStatementSchema,
  vehicle_title: vehicleTitleSchema,
  'tax.us.1099': tax1099Schema,
};

export function getSchemaForDocClass(docClass: DocClass): z.ZodType {
  return schemaMap[docClass] ?? genericSchema;
}

/**
 * Returns a JSON template string describing the expected fields for a doc class.
 * Used in AI extraction prompts.
 */
export function getExtractionPromptTemplate(docClass: DocClass): string {
  const templates: Partial<Record<DocClass, string>> = {
    'payStub.us': JSON.stringify({
      employer_name: 'string',
      employee_name: 'string',
      pay_period_start: 'YYYY-MM-DD',
      pay_period_end: 'YYYY-MM-DD',
      pay_date: 'YYYY-MM-DD',
      pay_frequency: 'weekly|biweekly|semimonthly|monthly',
      gross_pay: 0,
      federal_tax: 0,
      state_tax: 0,
      social_security: 0,
      medicare: 0,
      health_insurance: 0,
      retirement_401k: 0,
      other_deductions: [{ name: 'string', amount: 0 }],
      net_pay: 0,
      ytd_gross: 0,
      ytd_net: 0,
      hours_worked: 0,
      hourly_rate: 0,
    }, null, 2),

    'bankStatement.us.checking': JSON.stringify({
      institution_name: 'string',
      account_type: 'checking|savings',
      account_number_last4: 'string',
      statement_period_start: 'YYYY-MM-DD',
      statement_period_end: 'YYYY-MM-DD',
      beginning_balance: 0,
      ending_balance: 0,
      total_deposits: 0,
      total_withdrawals: 0,
      transactions: [{ date: 'YYYY-MM-DD', description: 'string', amount: 0, type: 'credit|debit' }],
    }, null, 2),

    'tax.us.w2': JSON.stringify({
      employer_name: 'string',
      employer_ein: 'string',
      employee_name: 'string',
      employee_ssn_last4: 'string (last 4 digits only)',
      tax_year: 'YYYY',
      wages: 0,
      federal_tax_withheld: 0,
      social_security_wages: 0,
      social_security_tax: 0,
      medicare_wages: 0,
      medicare_tax: 0,
      state: 'string',
      state_wages: 0,
      state_tax: 0,
    }, null, 2),

    'tax.us.1040': JSON.stringify({
      tax_year: 'YYYY',
      return_type: 'federal|state',
      filing_status: 'single|married_jointly|married_separately|head_of_household|qualifying_surviving_spouse',
      adjusted_gross_income: 0,
      taxable_income: 0,
      total_tax: 0,
      total_payments: 0,
      refund_amount: 0,
      amount_owed: 0,
    }, null, 2),

    creditCard: JSON.stringify({
      issuer: 'string (financial institution, not card network)',
      account_number_last4: 'string (last 4 digits only)',
      statement_period_start: 'YYYY-MM-DD',
      statement_period_end: 'YYYY-MM-DD',
      previous_balance: 0,
      payments: 0,
      new_charges: 0,
      ending_balance: 0,
      minimum_payment_due: 0,
      payment_due_date: 'YYYY-MM-DD',
      credit_limit: 0,
      available_credit: 0,
      cash_advances: 0,
    }, null, 2),

    profit_loss_statement: JSON.stringify({
      business_name: 'string',
      gross_revenue: 0,
      net_profit: 0,
      period_start: 'YYYY-MM-DD',
      period_end: 'YYYY-MM-DD',
      total_expenses: 0,
      owner_name: 'string',
      cost_of_goods_sold: 0,
      payroll_expenses: 0,
      rent_expense: 0,
      utilities: 0,
      other_expenses: [{ name: 'string', amount: 0 }],
    }, null, 2),

    retirement_account: JSON.stringify({
      institution_name: 'string',
      account_type: 'IRA|401k|403b|pension|other_retirement',
      ending_balance: 0,
      account_number_last4: 'string (last 4 digits only)',
      statement_period_end: 'YYYY-MM-DD',
      account_holder_name: 'string',
      employer_name: 'string',
    }, null, 2),

    collection_letter: JSON.stringify({
      collection_agency_name: 'string',
      amount_claimed: 0,
      original_creditor: 'string',
      account_number_last4: 'string (last 4 digits only)',
      letter_date: 'YYYY-MM-DD',
      debt_type: 'string (e.g. credit card, medical, auto loan)',
      references_lawsuit: false,
      references_judgment: false,
      judgment_amount: 0,
      court_name: 'string',
      collection_agency_address: 'string',
      phone: 'string',
    }, null, 2),

    legal_document: JSON.stringify({
      document_type: 'summons|complaint|judgment|garnishment_order|foreclosure_notice|other',
      plaintiff_name: 'string',
      defendant_name: 'string',
      case_number: 'string',
      court_name: 'string',
      court_address: 'string',
      filing_date: 'YYYY-MM-DD',
      case_type: 'string (e.g. debt collection, mortgage foreclosure)',
      amount_claimed: 0,
      judgment_amount: 0,
      garnishment_amount: 0,
      property_address: 'string',
    }, null, 2),

    vehicle_loan_statement: JSON.stringify({
      lender_name: 'string',
      current_balance: 0,
      account_number_last4: 'string (last 4 digits only)',
      interest_rate: 0.0699,
      monthly_payment: 0,
      vehicle_description: 'string (e.g. 2021 Toyota Camry)',
      loan_origination_date: 'YYYY-MM-DD',
      payoff_amount: 0,
      lender_address: 'string',
    }, null, 2),

    'mortgage.us': JSON.stringify({
      lender_name: 'string',
      current_balance: 0,
      loan_number: 'string (full loan number)',
      property_address: 'string',
      interest_rate: 0.065,
      monthly_payment: 0,
      statement_period_end: 'YYYY-MM-DD',
      loan_type: 'first_mortgage|second_mortgage|heloc|other',
      escrow_balance: 0,
      payoff_amount: 0,
      lender_address: 'string',
    }, null, 2),

    social_security_letter: JSON.stringify({
      monthly_benefit: 0,
      benefit_type: 'SSDI|SSI|retirement|survivor|other',
      effective_date: 'YYYY-MM-DD',
      recipient_name: 'string',
      net_monthly_benefit: 0,
      medicare_premium: 0,
      annual_benefit: 0,
    }, null, 2),

    idDocument: JSON.stringify({
      full_name: 'string',
      date_of_birth: 'YYYY-MM-DD',
      license_number: 'string',
      expiration_date: 'YYYY-MM-DD',
      address: 'string',
      state: 'string',
      sex: 'string',
    }, null, 2),

    social_security_card: JSON.stringify({
      full_name: 'string',
      ssn_last4: 'string (last 4 digits only)',
    }, null, 2),

    brokerage_statement: JSON.stringify({
      institution_name: 'string',
      account_holder_name: 'string',
      account_number_last4: 'string (last 4 digits only)',
      statement_period_start: 'YYYY-MM-DD',
      statement_period_end: 'YYYY-MM-DD',
      beginning_value: 0,
      ending_value: 0,
    }, null, 2),

    vehicle_title: JSON.stringify({
      title_number: 'string',
      vin: 'string',
      year: 'string',
      make: 'string',
      model: 'string',
      body_style: 'string',
      legal_owner_name: 'string',
      registered_owner_name: 'string',
      issue_date: 'YYYY-MM-DD',
      odometer_miles: 0,
      legal_owner_address: 'string',
      prior_title_state: 'string',
    }, null, 2),

    'tax.us.1099': JSON.stringify({
      form_variant: '1099-MISC|1099-NEC|1099-INT|1099-DIV|1099-R|1099-SSA|1099-G',
      payer_name: 'string',
      recipient_name: 'string',
      recipient_ssn_last4: 'string (last 4 digits only)',
      tax_year: 'YYYY',
      total_amount: 0,
      federal_tax_withheld: 0,
    }, null, 2),
  };

  // bank_statement_savings uses the same template as checking
  if (docClass === 'bankStatement.us.savings') {
    return templates['bankStatement.us.checking'] ?? '{}';
  }

  // Legacy retirement classes reuse the retirement_account template
  if (docClass === 'ira_statement' || docClass === '401k_statement') {
    return templates.retirement_account ?? '{}';
  }

  return templates[docClass] ?? '{}';
}

File: extractor/schemas.py
"""
Pydantic models mirroring the Zod schemas in server/src/services/extraction/schemas.ts.
Field names, types, and optionality must match exactly — the Node.js server consumes
this JSON without transformation.

Dates: YYYY-MM-DD strings.
Numbers: float, never string.
Optional fields: omitted when not found (not null or 0).
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from pydantic import BaseModel, ConfigDict


# ---- Extraction result (shared across all providers) -----------------------

@dataclass
class RuleExtractionResult:
    data: dict
    field_confidences: dict[str, float]
    warnings: list[str]
    confidence: float
    pages_analyzed: int = 0


# ---- Shared config ---------------------------------------------------------

class _Base(BaseModel):
    model_config = ConfigDict(extra="forbid")


# ---- Per-class data models -------------------------------------------------

class OtherDeduction(BaseModel):
    name: str
    amount: float


class PaystubData(_Base):
    employer_name: str
    employee_name: str | None = None
    pay_period_start: str | None = None
    pay_period_end: str | None = None
    pay_date: str | None = None
    pay_frequency: Literal["weekly", "biweekly", "semimonthly", "monthly"] | None = None
    gross_pay: float
    federal_tax: float | None = None
    state_tax: float | None = None
    social_security: float | None = None
    medicare: float | None = None
    health_insurance: float | None = None
    retirement_401k: float | None = None
    other_deductions: list[OtherDeduction] | None = None
    net_pay: float
    ytd_gross: float | None = None
    ytd_net: float | None = None
    hours_worked: float | None = None
    hourly_rate: float | None = None


class BankTransaction(BaseModel):
    date: str
    description: str
    amount: float
    type: Literal["credit", "debit"]


class BankStatementData(_Base):
    institution_name: str
    account_type: Literal["checking", "savings", "investment"] | None = None
    account_number_last4: str | None = None
    statement_period_start: str | None = None
    statement_period_end: str | None = None
    beginning_balance: float
    ending_balance: float
    total_deposits: float | None = None
    total_withdrawals: float | None = None
    transactions: list[BankTransaction] | None = None


class W2Data(_Base):
    employer_name: str
    employer_ein: str | None = None
    employee_name: str | None = None
    employee_ssn_last4: str | None = None
    tax_year: str | None = None
    wages: float
    federal_tax_withheld: float
    social_security_wages: float | None = None
    social_security_tax: float | None = None
    medicare_wages: float | None = None
    medicare_tax: float | None = None
    state: str | None = None
    state_wages: float | None = None
    state_tax: float | None = None


class TaxReturnData(_Base):
    tax_year: str
    return_type: Literal["federal", "state"] | None = None
    filing_status: Literal[
        "single",
        "married_jointly",
        "married_separately",
        "head_of_household",
        "qualifying_surviving_spouse",
    ] | None = None
    adjusted_gross_income: float
    taxable_income: float | None = None
    total_tax: float | None = None
    total_payments: float | None = None
    refund_amount: float | None = None
    amount_owed: float | None = None


class CreditCardData(_Base):
    issuer: str
    account_number_last4: str | None = None
    statement_period_start: str | None = None
    statement_period_end: str | None = None
    previous_balance: float
    payments: float | None = None
    new_charges: float | None = None
    ending_balance: float
    minimum_payment_due: float | None = None
    payment_due_date: str | None = None
    credit_limit: float | None = None
    available_credit: float | None = None
    cash_advances: float | None = None


class OtherExpense(BaseModel):
    name: str
    amount: float


class ProfitLossData(_Base):
    business_name: str
    gross_revenue: float
    net_profit: float
    period_start: str | None = None
    period_end: str | None = None
    total_expenses: float | None = None
    owner_name: str | None = None
    cost_of_goods_sold: float | None = None
    payroll_expenses: float | None = None
    rent_expense: float | None = None
    utilities: float | None = None
    other_expenses: list[OtherExpense] | None = None


class RetirementAccountData(_Base):
    institution_name: str
    account_type: Literal["IRA", "401k", "403b", "pension", "other_retirement"]
    ending_balance: float
    account_number_last4: str | None = None
    statement_period_end: str | None = None
    account_holder_name: str | None = None
    employer_name: str | None = None


class CollectionLetterData(_Base):
    collection_agency_name: str
    amount_claimed: float
    original_creditor: str | None = None
    account_number_last4: str | None = None
    letter_date: str | None = None
    debt_type: str | None = None
    references_lawsuit: bool | None = None
    references_judgment: bool | None = None
    judgment_amount: float | None = None
    court_name: str | None = None
    collection_agency_address: str | None = None
    phone: str | None = None


class LegalDocumentData(_Base):
    document_type: Literal[
        "summons", "complaint", "judgment", "garnishment_order", "foreclosure_notice", "other"
    ]
    plaintiff_name: str
    defendant_name: str | None = None
    case_number: str | None = None
    court_name: str | None = None
    court_address: str | None = None
    filing_date: str | None = None
    case_type: str | None = None
    amount_claimed: float | None = None
    judgment_amount: float | None = None
    garnishment_amount: float | None = None
    property_address: str | None = None


class VehicleLoanData(_Base):
    lender_name: str | None = None
    current_balance: float
    account_number_last4: str | None = None
    interest_rate: float | None = None
    monthly_payment: float | None = None
    vehicle_description: str | None = None
    loan_origination_date: str | None = None
    payoff_amount: float | None = None
    lender_address: str | None = None


class MortgageData(_Base):
    lender_name: str
    current_balance: float
    loan_number: str | None = None
    property_address: str | None = None
    interest_rate: float | None = None
    monthly_payment: float | None = None
    statement_period_end: str | None = None
    loan_type: Literal["first_mortgage", "second_mortgage", "heloc", "other"] | None = None
    escrow_balance: float | None = None
    payoff_amount: float | None = None
    lender_address: str | None = None


class MortgagePaymentData(_Base):
    lender_name: str
    payment_amount: float
    payment_date: str | None = None
    loan_number_last4: str | None = None
    confirmation_number: str | None = None


class SocialSecurityData(_Base):
    monthly_benefit: float
    benefit_type: Literal["SSDI", "SSI", "retirement", "survivor", "other"] | None = None
    effective_date: str | None = None
    recipient_name: str | None = None
    net_monthly_benefit: float | None = None
    medicare_premium: float | None = None
    annual_benefit: float | None = None


class DriversLicenseData(_Base):
    full_name: str
    date_of_birth: str | None = None   # YYYY-MM-DD
    license_number: str | None = None
    expiration_date: str | None = None  # YYYY-MM-DD
    address: str | None = None
    state: str | None = None
    sex: str | None = None


class SocialSecurityCardData(_Base):
    full_name: str
    # Full SSN is never stored — only last 4 digits if clearly visible
    ssn_last4: str | None = None


class BrokerageStatementData(_Base):
    institution_name: str
    account_holder_name: str | None = None
    account_number_last4: str | None = None
    statement_period_start: str | None = None
    statement_period_end: str | None = None
    beginning_value: float
    ending_value: float


class VehicleTitleData(_Base):
    vin: str | None = None
    year: str | None = None
    make: str | None = None
    model: str | None = None


class Tax1099Data(_Base):
    form_variant: str  # 1099-MISC, 1099-NEC, 1099-INT, 1099-DIV, 1099-R, 1099-SSA, 1099-G, etc.
    payer_name: str
    recipient_name: str | None = None
    recipient_ssn_last4: str | None = None
    tax_year: str | None = None
    total_amount: float
    federal_tax_withheld: float | None = None


# Union of all data types
ExtractionData = (
    PaystubData
    | BankStatementData
    | W2Data
    | TaxReturnData
    | CreditCardData
    | ProfitLossData
    | RetirementAccountData
    | CollectionLetterData
    | LegalDocumentData
    | VehicleLoanData
    | MortgageData
    | SocialSecurityData
    | DriversLicenseData
    | SocialSecurityCardData
    | BrokerageStatementData
    | VehicleTitleData
    | Tax1099Data
)

# All valid doc class strings (matches TypeScript DOC_CLASS_VALUES)
DOC_CLASSES = [
    "payStub.us",
    "bankStatement.us.checking",
    "bankStatement.us.savings",
    "tax.us.1040",
    "ira_statement",
    "401k_statement",
    "creditCard",
    "payroll_export",
    "tax.us.w2",
    "tax.us.1099",
    "other",
    "unclassified",
    "profit_loss_statement",
    "retirement_account",
    "collection_letter",
    "legal_document",
    "vehicle_loan_statement",
    "mortgage.us",
    "mortgage_payment",
    "social_security_letter",
    "idDocument",
    "social_security_card",
    "brokerage_statement",
    "vehicle_title",
]

# Map doc_class → data model (for validation)
DOC_CLASS_SCHEMA: dict[str, type[_Base]] = {
    "payStub.us": PaystubData,
    "bankStatement.us.checking": BankStatementData,
    "bankStatement.us.savings": BankStatementData,
    "tax.us.1040": TaxReturnData,
    "ira_statement": RetirementAccountData,
    "401k_statement": RetirementAccountData,
    "creditCard": CreditCardData,
    "tax.us.w2": W2Data,
    "profit_loss_statement": ProfitLossData,
    "retirement_account": RetirementAccountData,
    "collection_letter": CollectionLetterData,
    "legal_document": LegalDocumentData,
    "vehicle_loan_statement": VehicleLoanData,
    "mortgage.us": MortgageData,
    "mortgage_payment": MortgagePaymentData,
    "social_security_letter": SocialSecurityData,
    "idDocument": DriversLicenseData,
    "social_security_card": SocialSecurityCardData,
    "brokerage_statement": BrokerageStatementData,
    "vehicle_title": VehicleTitleData,
    "tax.us.1099": Tax1099Data,
}


# ---- API response model ----------------------------------------------------

class ExtractionResult(BaseModel):
    """The contract returned by POST /extract. Consumed by the Node.js server as-is."""
    doc_class: str
    classification_confidence: float
    classification_method: Literal["rule_engine", "ai"]
    extraction_method: str  # Provider name: "rule_engine", "ai_parse", "unclassified", etc.
    confidence: float
    data: dict  # Typed per doc_class — Node.js validates against its own Zod schema
    field_confidences: dict[str, float]
    warnings: list[str]
